// @pulumi/aws/lambda/runtimes.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DotnetCore2d1Runtime = "dotnetcore2.1";
exports.Go1dxRuntime = "go1.x";
exports.Java8Runtime = "java8";
exports.Java11Runtime = "java11";
exports.Ruby2d5Runtime = "ruby2.5";
exports.NodeJS10dXRuntime = "nodejs10.x";
exports.NodeJS12dXRuntime = "nodejs12.x";
exports.Python2d7Runtime = "python2.7";
exports.Python3d6Runtime = "python3.6";
exports.Python3d7Runtime = "python3.7";
exports.Python3d8Runtime = "python3.8";
exports.CustomRuntime = "provided";
/** @deprecated No longer supported. New lambda functions created using this runtime will fail */
exports.NodeJSRuntime = "nodejs";
/** @deprecated No longer supported. New lambda functions created using this runtime will fail */
exports.NodeJS4d3EdgeRuntime = "nodejs4.3-edge";
/** @deprecated No longer supported. New lambda functions created using this runtime will fail */
exports.NodeJS4d3Runtime = "nodejs4.3";
/** @deprecated No longer supported. New lambda functions created using this runtime will fail */
exports.NodeJS6d10Runtime = "nodejs6.10";
/** @deprecated No longer supported. New lambda functions created using this runtime will fail */
exports.DotnetCore1d0Runtime = "dotnetcore1.0";
/** @deprecated No longer supported. New lambda functions created using this runtime will fail */
exports.DotnetCore2d0Runtime = "dotnetcore2.0";
/** @deprecated New lambdas created using this runtime will succeed until December 31, 2019, at which point they may start to fail */
exports.NodeJS8d10Runtime = "nodejs8.10";
