// @pulumi/aws/appsync/function.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an AppSync Function.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const testGraphQLApi = new aws.appsync.GraphQLApi("test", {
 *     authenticationType: "API_KEY",
 *     schema: `type Mutation {
 *     putPost(id: ID!, title: String!): Post
 * }
 *
 * type Post {
 *     id: ID!
 *     title: String!
 * }
 *
 * type Query {
 *     singlePost(id: ID!): Post
 * }
 *
 * schema {
 *     query: Query
 *     mutation: Mutation
 * }
 * `,
 * });
 * const testDataSource = new aws.appsync.DataSource("test", {
 *     apiId: testGraphQLApi.id,
 *     httpConfig: {
 *         endpoint: "http://example.com",
 *     },
 *     type: "HTTP",
 * });
 * const testFunction = new aws.appsync.Function("test", {
 *     apiId: testGraphQLApi.id,
 *     dataSource: testDataSource.name,
 *     name: "tfExample",
 *     requestMappingTemplate: `{
 *     "version": "2018-05-29",
 *     "method": "GET",
 *     "resourcePath": "/",
 *     "params":{
 *         "headers": $utils.http.copyheaders($ctx.request.headers)
 *     }
 * }
 * `,
 *     responseMappingTemplate: `#if($ctx.result.statusCode == 200)
 *     $ctx.result.body
 * #else
 *     $utils.appendError($ctx.result.body, $ctx.result.statusCode)
 * #end
 * `,
 * });
 * ```
 *
 * > This content is derived from https://github.com/terraform-providers/terraform-provider-aws/blob/master/website/docs/r/appsync_function.html.markdown.
 */
class Function extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["apiId"] = state ? state.apiId : undefined;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["dataSource"] = state ? state.dataSource : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["functionId"] = state ? state.functionId : undefined;
            inputs["functionVersion"] = state ? state.functionVersion : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["requestMappingTemplate"] = state ? state.requestMappingTemplate : undefined;
            inputs["responseMappingTemplate"] = state ? state.responseMappingTemplate : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.apiId === undefined) {
                throw new Error("Missing required property 'apiId'");
            }
            if (!args || args.dataSource === undefined) {
                throw new Error("Missing required property 'dataSource'");
            }
            if (!args || args.requestMappingTemplate === undefined) {
                throw new Error("Missing required property 'requestMappingTemplate'");
            }
            if (!args || args.responseMappingTemplate === undefined) {
                throw new Error("Missing required property 'responseMappingTemplate'");
            }
            inputs["apiId"] = args ? args.apiId : undefined;
            inputs["dataSource"] = args ? args.dataSource : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["functionVersion"] = args ? args.functionVersion : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["requestMappingTemplate"] = args ? args.requestMappingTemplate : undefined;
            inputs["responseMappingTemplate"] = args ? args.responseMappingTemplate : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["functionId"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Function.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Function resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     */
    static get(name, id, state, opts) {
        return new Function(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Function.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Function.__pulumiType;
    }
}
exports.Function = Function;
/** @internal */
Function.__pulumiType = 'aws:appsync/function:Function';
