// @pulumi/aws/codestarnotifications/notificationRule.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a CodeStar Notifications Rule.
 *
 * > This content is derived from https://github.com/terraform-providers/terraform-provider-aws/blob/master/website/docs/r/codestarnotifications_notification_rule.html.markdown.
 */
class NotificationRule extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["detailType"] = state ? state.detailType : undefined;
            inputs["eventTypeIds"] = state ? state.eventTypeIds : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["resource"] = state ? state.resource : undefined;
            inputs["status"] = state ? state.status : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["targets"] = state ? state.targets : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.detailType === undefined) {
                throw new Error("Missing required property 'detailType'");
            }
            if (!args || args.eventTypeIds === undefined) {
                throw new Error("Missing required property 'eventTypeIds'");
            }
            if (!args || args.resource === undefined) {
                throw new Error("Missing required property 'resource'");
            }
            inputs["detailType"] = args ? args.detailType : undefined;
            inputs["eventTypeIds"] = args ? args.eventTypeIds : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["resource"] = args ? args.resource : undefined;
            inputs["status"] = args ? args.status : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["targets"] = args ? args.targets : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(NotificationRule.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing NotificationRule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     */
    static get(name, id, state, opts) {
        return new NotificationRule(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of NotificationRule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === NotificationRule.__pulumiType;
    }
}
exports.NotificationRule = NotificationRule;
/** @internal */
NotificationRule.__pulumiType = 'aws:codestarnotifications/notificationRule:NotificationRule';
