// @pulumi/aws/ram/resourceShareAccepter.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * > This content is derived from https://github.com/terraform-providers/terraform-provider-aws/blob/master/website/docs/r/ram_resource_share_accepter.html.markdown.
 */
class ResourceShareAccepter extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["invitationArn"] = state ? state.invitationArn : undefined;
            inputs["receiverAccountId"] = state ? state.receiverAccountId : undefined;
            inputs["resources"] = state ? state.resources : undefined;
            inputs["senderAccountId"] = state ? state.senderAccountId : undefined;
            inputs["shareArn"] = state ? state.shareArn : undefined;
            inputs["shareId"] = state ? state.shareId : undefined;
            inputs["shareName"] = state ? state.shareName : undefined;
            inputs["status"] = state ? state.status : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.shareArn === undefined) {
                throw new Error("Missing required property 'shareArn'");
            }
            inputs["shareArn"] = args ? args.shareArn : undefined;
            inputs["invitationArn"] = undefined /*out*/;
            inputs["receiverAccountId"] = undefined /*out*/;
            inputs["resources"] = undefined /*out*/;
            inputs["senderAccountId"] = undefined /*out*/;
            inputs["shareId"] = undefined /*out*/;
            inputs["shareName"] = undefined /*out*/;
            inputs["status"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(ResourceShareAccepter.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ResourceShareAccepter resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     */
    static get(name, id, state, opts) {
        return new ResourceShareAccepter(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResourceShareAccepter.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResourceShareAccepter.__pulumiType;
    }
}
exports.ResourceShareAccepter = ResourceShareAccepter;
/** @internal */
ResourceShareAccepter.__pulumiType = 'aws:ram/resourceShareAccepter:ResourceShareAccepter';
