// @pulumi/docker/config/vars.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
let __config = new pulumi.Config("docker");
/**
 * PEM-encoded content of Docker host CA certificate
 */
exports.caMaterial = __config.get("caMaterial") || utilities.getEnv("DOCKER_CA_MATERIAL");
/**
 * PEM-encoded content of Docker client certificate
 */
exports.certMaterial = __config.get("certMaterial") || utilities.getEnv("DOCKER_CERT_MATERIAL");
/**
 * Path to directory with Docker TLS config
 */
exports.certPath = __config.get("certPath") || utilities.getEnv("DOCKER_CERT_PATH");
/**
 * The Docker daemon address
 */
exports.host = __config.get("host") || (utilities.getEnv("DOCKER_HOST") || "unix:///var/run/docker.sock");
/**
 * PEM-encoded content of Docker client private key
 */
exports.keyMaterial = __config.get("keyMaterial") || utilities.getEnv("DOCKER_KEY_MATERIAL");
exports.registryAuth = __config.getObject("registryAuth");
