// @pulumi/aws/rds/globalCluster.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages a RDS Global Cluster, which is an Aurora global database spread across multiple regions. The global database contains a single primary cluster with read-write capability, and a read-only secondary cluster that receives data from the primary cluster through high-speed replication performed by the Aurora storage subsystem.
 *
 * More information about Aurora global databases can be found in the [Aurora User Guide](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database.html#aurora-global-database-creating).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const primary = new aws.Provider("primary", {
 *     region: "us-east-2",
 * });
 * const secondary = new aws.Provider("secondary", {
 *     region: "us-west-2",
 * });
 * const example = new aws.rds.GlobalCluster("example", {
 *     globalClusterIdentifier: "example",
 * }, {provider: primary});
 * const primaryCluster = new aws.rds.Cluster("primary", {
 *     // ... other configuration ...
 *     engineMode: "global",
 *     globalClusterIdentifier: example.id,
 * }, {provider: primary});
 * const primaryClusterInstance = new aws.rds.ClusterInstance("primary", {
 *     // ... other configuration ...
 *     clusterIdentifier: primaryCluster.id,
 * }, {provider: primary});
 * const secondaryCluster = new aws.rds.Cluster("secondary", {
 *     // ... other configuration ...
 *     engineMode: "global",
 *     globalClusterIdentifier: example.id,
 * }, {provider: secondary,dependsOn: [primaryClusterInstance]});
 * const secondaryClusterInstance = new aws.rds.ClusterInstance("secondary", {
 *     // ... other configuration ...
 *     clusterIdentifier: secondaryCluster.id,
 * }, {provider: secondary});
 * ```
 *
 * > This content is derived from https://github.com/terraform-providers/terraform-provider-aws/blob/master/website/docs/r/rds_global_cluster.html.markdown.
 */
class GlobalCluster extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["databaseName"] = state ? state.databaseName : undefined;
            inputs["deletionProtection"] = state ? state.deletionProtection : undefined;
            inputs["engine"] = state ? state.engine : undefined;
            inputs["engineVersion"] = state ? state.engineVersion : undefined;
            inputs["globalClusterIdentifier"] = state ? state.globalClusterIdentifier : undefined;
            inputs["globalClusterResourceId"] = state ? state.globalClusterResourceId : undefined;
            inputs["storageEncrypted"] = state ? state.storageEncrypted : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.globalClusterIdentifier === undefined) {
                throw new Error("Missing required property 'globalClusterIdentifier'");
            }
            inputs["databaseName"] = args ? args.databaseName : undefined;
            inputs["deletionProtection"] = args ? args.deletionProtection : undefined;
            inputs["engine"] = args ? args.engine : undefined;
            inputs["engineVersion"] = args ? args.engineVersion : undefined;
            inputs["globalClusterIdentifier"] = args ? args.globalClusterIdentifier : undefined;
            inputs["storageEncrypted"] = args ? args.storageEncrypted : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["globalClusterResourceId"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(GlobalCluster.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing GlobalCluster resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     */
    static get(name, id, state, opts) {
        return new GlobalCluster(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of GlobalCluster.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === GlobalCluster.__pulumiType;
    }
}
exports.GlobalCluster = GlobalCluster;
/** @internal */
GlobalCluster.__pulumiType = 'aws:rds/globalCluster:GlobalCluster';
