// @pulumi/aws/region.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.APEast1Region = "ap-east-1";
exports.APNortheast1Region = "ap-northeast-1";
exports.APNortheast2Region = "ap-northeast-2";
exports.APSouth1Region = "ap-south-1";
exports.APSouthEast2Region = "ap-southeast-2";
exports.APSoutheast1Region = "ap-southeast-1";
exports.CACentralRegion = "ca-central-1";
exports.CNNorth1Region = "cn-north-1";
exports.CNNorthWest1Region = "cn-northwest-1";
exports.EUCentral1Region = "eu-central-1";
exports.EUNorth1Region = "eu-north-1";
exports.EUWest1Region = "eu-west-1";
exports.EUWest2Region = "eu-west-2";
exports.EUWest3Region = "eu-west-3";
exports.MESouth1Region = "me-south-1";
exports.SAEast1Region = "sa-east-1";
exports.USEast1Region = "us-east-1";
exports.USEast2Region = "us-east-2";
exports.USWest1Region = "us-west-1";
exports.USWest2Region = "us-west-2";
