// @pulumi/aws/iam/principals.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
var Principals;
(function (Principals) {
    /**
     * Service Principal for Amazon Certificate Manager
     */
    Principals.AcmServicePrincipal = { Service: "acm.amazonaws.com" };
    /**
     * Service Principal for API Gateway
     */
    Principals.ApiGatewayPrincipal = { Service: "apigateway.amazonaws.com" };
    /**
     * Service Principal for Athena
     */
    Principals.AthenaPrincipal = { Service: "athena.amazonaws.com" };
    /**
     * Service Principal for Autoscaling
     */
    Principals.AutoscalingPrincipal = { Service: "autoscaling.amazonaws.com" };
    /**
     * Service Principal for Cloud Directory
     */
    Principals.CloudDirectoryPrincipal = { Service: "clouddirectory.amazonaws.com" };
    /**
     * Service Principal for Cloudformation
     */
    Principals.CloudformationPrincipal = { Service: "cloudformation.amazonaws.com" };
    /**
     * Service Principal for Cloudfront
     */
    Principals.CloudfrontPrincipal = { Service: "cloudfront.amazonaws.com" };
    /**
     * Service Principal for Cloud Search
     */
    Principals.CloudSearchPrincipal = { Service: "cloudsearch.amazonaws.com" };
    /**
     * Service Principal for Cloudtrail
     */
    Principals.CloudtrailPrincipal = { Service: "cloudtrail.amazonaws.com" };
    /**
     * Service Principal for CodeCommit
     */
    Principals.CodeCommitPrincipal = { Service: "codecommit.amazonaws.com" };
    /**
     * Service Principal for CodeBuild
     */
    Principals.CodeBuildPrincipal = { Service: "codebuild.amazonaws.com" };
    /**
     * Service Principal for CodeDeploy
     */
    Principals.CodeDeployPrincipal = { Service: "codedeploy.amazonaws.com" };
    /**
     * Service Principal for CodePipeline
     */
    Principals.CodePipelinePrincipal = { Service: "codepipeline.amazonaws.com" };
    /**
     * Service Principal for EC2 Config Service
     */
    Principals.ConfigPrincipal = { Service: "config.amazonaws.com" };
    /**
     * Service Principal for Data Pipeline
     */
    Principals.DataPipelinePrincipal = { Service: "datapipeline.amazonaws.com" };
    /**
     * Service Principal for DirectConnect
     */
    Principals.DirectConnectPrincipal = { Service: "directconnect.amazonaws.com" };
    /**
     * Service Principal for Directory Services
     */
    Principals.DirectoryServicesPrincipal = { Service: "ds.amazonaws.com" };
    /**
     * Service Principal for DynamoDB
     */
    Principals.DynamoDbPrincipal = { Service: "dynamodb.amazonaws.com" };
    /**
     * Service Principal for EC2
     */
    Principals.Ec2Principal = { Service: "ec2.amazonaws.com" };
    /**
     * Service Principal for Elastic Container Registry
     */
    Principals.EcrPrincipal = { Service: "ecr.amazonaws.com" };
    /**
     * Service Principal for Elastic Container Service
     */
    Principals.EcsPrincipal = { Service: "ecs.amazonaws.com" };
    /**
     * Service Principal for Edge Lambda
     */
    Principals.EdgeLambdaPrincipal = { Service: "edgelambda.amazonaws.com" };
    /**
     * Service Principal for Elasticache
     */
    Principals.ElasticachePrincipal = { Service: "elasticache.amazonaws.com" };
    /**
     * Service Principal for Elastic Beanstalk
     */
    Principals.ElasticBeanstalkPrincipal = { Service: "elasticbeanstalk.amazonaws.com" };
    /**
     * Service Principal for Elastic File System
     */
    Principals.ElasticFileSystemPrincipal = { Service: "elasticfilesystem.amazonaws.com" };
    /**
     * Service Principal for Elastic Load Balancing
     */
    Principals.ElasticLoadBalancingPrincipal = { Service: "elasticloadbalancing.amazonaws.com" };
    /**
     * Service Principal for Elastic MapReduce
     */
    Principals.ElasticMapReducePrincipal = { Service: "elasticmapreduce.amazonaws.com" };
    /**
     * Service Principal for Events
     */
    Principals.EventsPrincipal = { Service: "events.amazonaws.com" };
    /**
     * Service Principal for Health
     */
    Principals.HealthPrincipal = { Service: "health.amazonaws.com" };
    /**
     * Service Principal for IAM
     */
    Principals.IamPrincipal = { Service: "iam.amazonaws.com" };
    /**
     * Service Principal for AWS Inspector
     */
    Principals.InspectorPrincipal = { Service: "inspector.amazonaws.com" };
    /**
     * Service Principal for Kinesis
     */
    Principals.KinesisPrincipal = { Service: "kinesis.amazonaws.com" };
    /**
     * Service Principal for Key Mangaement Service
     */
    Principals.KmsPrincipal = { Service: "kms.amazonaws.com" };
    /**
     * Service Principal for Lambda
     */
    Principals.LambdaPrincipal = { Service: "lambda.amazonaws.com" };
    /**
     * Service Principal for Lightsail
     */
    Principals.LightsailPrincipal = { Service: "lightsail.amazonaws.com" };
    /**
     * Service Principal for Cloudwatch Logs
     */
    Principals.LogsPrincipal = { Service: "logs.amazonaws.com" };
    /**
     * Service Principal for Cloudwatch Monitoring
     */
    Principals.MonitoringPrincipal = { Service: "monitoring.amazonaws.com" };
    /**
     * Service Principal for Opsworks
     */
    Principals.OpsworksPrincipal = { Service: "opsworks.amazonaws.com" };
    /**
     * Service Principal for Organizations
     */
    Principals.OrganizationsPrincipal = { Service: "organizations.amazonaws.com" };
    /**
     * Service Principal for Relational Database Service
     */
    Principals.RdsPrincipal = { Service: "rds.amazonaws.com" };
    /**
     * Service Principal for Redshift
     */
    Principals.RedshiftPrincipal = { Service: "redshift.amazonaws.com" };
    /**
     * Service Principal for Route 53
     */
    Principals.Route53Principal = { Service: "route53.amazonaws.com" };
    /**
     * Service Principal for S3
     */
    Principals.S3Principal = { Service: "s3.amazonaws.com" };
    /**
     * Service Principal for Service Catalog
     */
    Principals.ServiceCatalogPrincipal = { Service: "servicecatalog.amazonaws.com" };
    /**
     * Service Principal for Simple Email Service
     */
    Principals.SesPrincipal = { Service: "ses.amazonaws.com" };
    /**
     * Service Principal for Signin Service
     */
    Principals.SigninPrincipal = { Service: "signin.amazonaws.com" };
    /**
     * Service Principal for Simple Notification Service
     */
    Principals.SnsPrincipal = { Service: "sns.amazonaws.com" };
    /**
     * Service Principal for Simple Queue Service
     */
    Principals.SqsPrincipal = { Service: "sqs.amazonaws.com" };
    /**
     * Service Principal for Systems Manager
     */
    Principals.SsmPrincipal = { Service: "ssm.amazonaws.com" };
    /**
     * Service Principal for Storage Gateway
     */
    Principals.StorageGatewayPrincipal = { Service: "storagegateway.amazonaws.com" };
    /**
     * Service Principal for Security Token Service
     */
    Principals.StsPrincipal = { Service: "sts.amazonaws.com" };
    /**
     * Service Principal for AWS Support
     */
    Principals.SupportPrincipal = { Service: "support.amazonaws.com" };
    /**
     * Service Principal for VM Import/Export
     */
    Principals.VmiePrincipal = { Service: "vmie.amazonaws.com" };
    /**
     * Service Principal for VPC Flow Logs
     */
    Principals.VpcFlowLogsPrincipal = { Service: "vpc-flow-logs.amazonaws.com" };
    /**
     * Service Principal for Web Application Firewall
     */
    Principals.WafPrincipal = { Service: "waf.amazonaws.com" };
    /**
     * Service Principal for WorkDocs
     */
    Principals.WorkDocsPrincipal = { Service: "workdocs.amazonaws.com" };
    /**
     * Service Principal for Workspaces
     */
    Principals.WorkspacesPrincipal = { Service: "workspaces.amazonaws.com" };
})(Principals = exports.Principals || (exports.Principals = {}));
/**
 * @deprecated
 * Service Principal for Amazon Certificate Manager
 */
exports.AcmServicePrincipal = { Service: "acm.amazonaws.com" };
/**
 * @deprecated
 * Service Principal for API Gateway
 */
exports.ApiGatewayPrincipal = { Service: "apigateway.amazonaws.com" };
/**
 * @deprecated
 * Service Principal for Athena
 */
exports.AthenaPrincipal = { Service: "athena.amazonaws.com" };
/**
 * @deprecated
 * Service Principal for Autoscaling
 */
exports.AutoscalingPrincipal = { Service: "autoscaling.amazonaws.com" };
/**
 * @deprecated
 * Service Principal for Cloud Directory
 */
exports.CloudDirectoryPrincipal = { Service: "clouddirectory.amazonaws.com" };
/**
 * @deprecated
 * Service Principal for Cloudformation
 */
exports.CloudformationPrincipal = { Service: "cloudformation.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Cloudfront
 */
exports.CloudfrontPrincipal = { Service: "cloudfront.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Cloud Search
 */
exports.CloudSearchPrincipal = { Service: "cloudsearch.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Cloudtrail
 */
exports.CloudtrailPrincipal = { Service: "cloudtrail.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for CodeCommit
 */
exports.CodeCommitPrincipal = { Service: "codecommit.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for CodeDeploy
 */
exports.CodeDeployPrincipal = { Service: "codedeploy.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for CodePipeline
 */
exports.CodePipelinePrincipal = { Service: "codepipeline.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for EC2 Config Service
 */
exports.ConfigPrincipal = { Service: "config.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Data Pipeline
 */
exports.DataPipelinePrincipal = { Service: "datapipeline.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for DirectConnect
 */
exports.DirectConnectPrincipal = { Service: "directconnect.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Directory Services
 */
exports.DirectoryServicesPrincipal = { Service: "ds.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for DynamoDB
 */
exports.DynamoDbPrincipal = { Service: "dynamodb.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for EC2
 */
exports.Ec2Principal = { Service: "ec2.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elastic Container Registry
 */
exports.EcrPrincipal = { Service: "ecr.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elastic Container Service
 */
exports.EcsPrincipal = { Service: "ecs.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Edge Lambda
 */
exports.EdgeLambdaPrincipal = { Service: "edgelambda.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elasticache
 */
exports.ElasticachePrincipal = { Service: "elasticache.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elastic Beanstalk
 */
exports.ElasticBeanstalkPrincipal = { Service: "elasticbeanstalk.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elastic File System
 */
exports.ElasticFileSystemPrincipal = { Service: "elasticfilesystem.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elastic Load Balancing
 */
exports.ElasticLoadBalancingPrincipal = { Service: "elasticloadbalancing.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Elastic MapReduce
 */
exports.ElasticMapReducePrincipal = { Service: "elasticmapreduce.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Events
 */
exports.EventsPrincipal = { Service: "events.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Health
 */
exports.HealthPrincipal = { Service: "health.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for IAM
 */
exports.IamPrincipal = { Service: "iam.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for AWS Inspector
 */
exports.InspectorPrincipal = { Service: "inspector.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Kinesis
 */
exports.KinesisPrincipal = { Service: "kinesis.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Key Mangaement Service
 */
exports.KmsPrincipal = { Service: "kms.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Lambda
 */
exports.LambdaPrincipal = { Service: "lambda.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Lightsail
 */
exports.LightsailPrincipal = { Service: "lightsail.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Cloudwatch Logs
 */
exports.LogsPrincipal = { Service: "logs.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Cloudwatch Monitoring
 */
exports.MonitoringPrincipal = { Service: "monitoring.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Opsworks
 */
exports.OpsworksPrincipal = { Service: "opsworks.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Organizations
 */
exports.OrganizationsPrincipal = { Service: "organizations.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Relational Database Service
 */
exports.RdsPrincipal = { Service: "rds.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Redshift
 */
exports.RedshiftPrincipal = { Service: "redshift.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Route 53
 */
exports.Route53Principal = { Service: "route53.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for S3
 */
exports.S3Principal = { Service: "s3.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Service Catalog
 */
exports.ServiceCatalogPrincipal = { Service: "servicecatalog.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Simple Email Service
 */
exports.SesPrincipal = { Service: "ses.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Signin Service
 */
exports.SigninPrincipal = { Service: "signin.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Simple Notification Service
 */
exports.SnsPrincipal = { Service: "sns.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Simple Queue Service
 */
exports.SqsPrincipal = { Service: "sqs.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Systems Manager
 */
exports.SsmPrincipal = { Service: "ssm.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Storage Gateway
 */
exports.StorageGatewayPrincipal = { Service: "storagegateway.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Security Token Service
 */
exports.StsPrincipal = { Service: "sts.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for AWS Support
 */
exports.SupportPrincipal = { Service: "support.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for VM Import/Export
 */
exports.VmiePrincipal = { Service: "vmie.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for VPC Flow Logs
 */
exports.VpcFlowLogsPrincipal = { Service: "vpc-flow-logs.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Web Application Firewall
 */
exports.WafPrincipal = { Service: "waf.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for WorkDocs
 */
exports.WorkDocsPrincipal = { Service: "workdocs.amazonaws.com" };
/**
 * @deprecated Use the equivalent constant in module `Principals` instead
 * Service Principal for Workspaces
 */
exports.WorkspacesPrincipal = { Service: "workspaces.amazonaws.com" };
