// @pulumi/aws/iam/samlProvider.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an IAM SAML provider.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 * import * as fs from "fs";
 *
 * const defaultSamlProvider = new aws.iam.SamlProvider("default", {
 *     samlMetadataDocument: fs.readFileSync("saml-metadata.xml", "utf-8"),
 * });
 * ```
 *
 * > This content is derived from https://github.com/terraform-providers/terraform-provider-aws/blob/master/website/docs/r/iam_saml_provider.html.markdown.
 */
class SamlProvider extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["samlMetadataDocument"] = state ? state.samlMetadataDocument : undefined;
            inputs["validUntil"] = state ? state.validUntil : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.samlMetadataDocument === undefined) {
                throw new Error("Missing required property 'samlMetadataDocument'");
            }
            inputs["name"] = args ? args.name : undefined;
            inputs["samlMetadataDocument"] = args ? args.samlMetadataDocument : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["validUntil"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(SamlProvider.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing SamlProvider resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     */
    static get(name, id, state, opts) {
        return new SamlProvider(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of SamlProvider.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SamlProvider.__pulumiType;
    }
}
exports.SamlProvider = SamlProvider;
/** @internal */
SamlProvider.__pulumiType = 'aws:iam/samlProvider:SamlProvider';
