// @pulumi/aws/s3/cannedAcl.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
// This file contains two categories of exports:
//
//     1) A union type, CannedAcl, that accepts any valid canned ACL type
//     2) Individual constants for each such canned ACL type
//
// These give a better developer experience and are just sugared strings.
exports.PrivateAcl = "private";
exports.PublicReadAcl = "public-read";
exports.PublicReadWriteAcl = "public-read-write";
exports.AwsExecReadAcl = "aws-exec-read";
exports.AuthenticatedReadAcl = "authenticated-read";
exports.BucketOwnerReadAcl = "bucket-owner-read";
exports.BucketOwnerFullControlAcl = "bucket-owner-full-control";
exports.LogDeliveryWriteAcl = "log-delivery-write";
